File: src/main/java/com/demonwav/mcdev/util/McPsiUtil.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 *
 * MIT License
 */

package com.demonwav.mcdev.util;

import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Lists;
import com.intellij.navigation.AnonymousElementProvider;
import com.intellij.openapi.extensions.Extensions;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Pair;
import com.intellij.psi.JavaPsiFacade;
import com.intellij.psi.JavaTokenType;
import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiJavaCodeReferenceElement;
import com.intellij.psi.PsiKeyword;
import com.intellij.psi.PsiMember;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiModifier;
import com.intellij.psi.PsiModifierList;
import com.intellij.psi.PsiModifierListOwner;
import com.intellij.psi.PsiReferenceList;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public final class McPsiUtil {
    private McPsiUtil() {}

    @Nullable
    @Contract(value = "null -> null", pure = true)
    public static PsiClass getClassOfElement(@Nullable PsiElement element) {
        if (element == null) {
            return null;
        }

        if (element instanceof PsiClass) {
            return (PsiClass) element;
        }

        while (element.getParent() != null) {

            if (element.getParent() instanceof PsiClass) {
                return (PsiClass) element.getParent();
            }

            if (element.getParent() instanceof PsiFile || element.getParent() instanceof PsiDirectory) {
                return null;
            }

            element = element.getParent();
        }
        return null;
    }

    public static boolean extendsOrImplementsClass(@NotNull PsiClass psiClass, @NotNull String qualifiedClassName) {
        final PsiClass[] supers = psiClass.getSupers();
        for (PsiClass aSuper : supers) {
            if (qualifiedClassName.equals(aSuper.getQualifiedName())) {
                return true;
            }
            if (extendsOrImplementsClass(aSuper, qualifiedClassName)) {
                return true;
            }
        }
        return false;
    }

    public static void addImplements(@NotNull PsiClass psiClass, @NotNull String qualifiedClassName, @NotNull Project project) {
        PsiReferenceList referenceList = psiClass.getImplementsList();
        PsiClass listenerClass = JavaPsiFacade.getInstance(project).findClass(qualifiedClassName, GlobalSearchScope.allScope(project));
        if (listenerClass != null) {
            PsiJavaCodeReferenceElement element = JavaPsiFacade.getElementFactory(project).createClassReferenceElement(listenerClass);
            if (referenceList != null) {
                referenceList.add(element);
            } else {
                PsiReferenceList list = JavaPsiFacade.getElementFactory(project).createReferenceList(new PsiJavaCodeReferenceElement[] {element});
                psiClass.add(list);
            }
        }
    }

    private static final ImmutableSet<String> METHOD_ACCESS_MODIFIERS = ImmutableSet.<String>builder()
            .add(PsiModifier.PUBLIC)
            .add(PsiModifier.PROTECTED)
            .add(PsiModifier.PACKAGE_LOCAL)
            .add(PsiModifier.PRIVATE)
            .build();

    public static String getAccessModifier(PsiMember member) {
        return METHOD_ACCESS_MODIFIERS.stream()
                .filter(member::hasModifierProperty)
                .findFirst()
                .orElse(PsiModifier.PUBLIC);
    }

    public static IElementType getMethodAccessType(PsiMethod method) {
        for (PsiElement modifier : method.getModifierList().getChildren()) {
            if (modifier instanceof PsiKeyword) {
                final IElementType tokenType = ((PsiKeyword) modifier).getTokenType();

            }
        }
        return JavaTokenType.PUBLIC_KEYWORD;
    }

    @Nullable
    public static PsiAnnotation getAnnotation(@Nullable PsiModifierListOwner owner, @NotNull String annotationName) {
        if (owner == null) {
            return null;
        }

        final PsiModifierList list = owner.getModifierList();
        if (list == null) {
            return null;
        }

        return list.findAnnotation(annotationName);
    }

    @Nullable
    @Contract(value = "null -> null", pure = true)
    public static Pair<String, PsiClass> getNameOfClass(@Nullable PsiClass psiClass) {
        if (psiClass == null) {
            return null;
        }

        if (psiClass.getContainingClass() == null) {
            //noinspection ConstantConditions
            return Pair.create("", psiClass);
        }

        final List<String> innerStrings = Lists.newArrayList();
        PsiClass baseClass = psiClass;
        while (psiClass != null) {
            baseClass = psiClass;
            if (psiClass.getName() == null) {
                // anon class
                PsiElement[] anonymousClasses = null;
                for (AnonymousElementProvider provider : Extensions.getExtensions(AnonymousElementProvider.EP_NAME)) {
                    //noinspection ConstantConditions
                    anonymousClasses = provider.getAnonymousElements(psiClass.getContainingClass());
                    if (anonymousClasses.length > 0) {
                        break;
                    }
                }

                if (anonymousClasses == null) {
                    // We couldn't build the proper string, so don't return anything at all
                    return null;
                }

                for (int i = 0; i < anonymousClasses.length; i++) {
                    if (anonymousClasses[i] == psiClass) {
                        innerStrings.add(String.valueOf(i + 1));
                        break;
                    }
                }
            } else {
                innerStrings.add(psiClass.getName());
                psiClass = psiClass.getContainingClass();
            }
        }

        // We started from the bottom and went up, so reverse it
        Collections.reverse(innerStrings);
        // Skip the base class, we are giving the base PsiClass so the user can do with it what they want
        return Pair.create("$" + innerStrings.stream().skip(1).collect(Collectors.joining("$")), baseClass);
    }
}

File: src/main/kotlin/com/demonwav/mcdev/util/McPsiUtil.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2017 minecraft-dev
 *
 * MIT License
 */

@file:JvmName("McPsiUtil")
package com.demonwav.mcdev.util

import com.intellij.psi.JavaPsiFacade
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementResolveResult
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiJavaCodeReferenceElement
import com.intellij.psi.PsiMember
import com.intellij.psi.PsiMethod
import com.intellij.psi.PsiParameter
import com.intellij.psi.PsiParameterList
import com.intellij.psi.PsiReference
import com.intellij.psi.PsiType
import com.intellij.psi.ResolveResult
import com.intellij.psi.filters.ElementFilter
import com.intellij.psi.util.TypeConversionUtil
import com.intellij.refactoring.changeSignature.ChangeSignatureUtil
import org.jetbrains.annotations.Contract
import java.util.stream.Stream

// Parent

@Contract(pure = true)
fun PsiElement.findContainingClass(): PsiClass? = findParent(resolveReferences = false)

@Contract(pure = true)
fun PsiElement.findReferencedClass(): PsiClass? = findParent(resolveReferences = true)

@Contract(pure = true)
fun PsiElement.findReferencedMember(): PsiMember? = findParent({ it is PsiClass }, resolveReferences = true)

@Contract(pure = true)
fun PsiElement.findContainingMethod(): PsiMethod? = findParent({ it is PsiClass }, resolveReferences = false)

@Contract(pure = true)
fun PsiElement.findJavaCodeReferenceElement(): PsiJavaCodeReferenceElement? {
    return findParent({ it is PsiMethod || it is PsiClass }, resolveReferences = false)
}

@Contract(pure = true)
private inline fun <reified T : PsiElement> PsiElement.findParent(resolveReferences: Boolean): T? {
    return findParent({false}, resolveReferences)
}

@Contract(pure = true)
private inline fun <reified T : PsiElement> PsiElement.findParent(stop: (PsiElement) -> Boolean, resolveReferences: Boolean): T? {
    var el: PsiElement = this

    while (true) {
        if (resolveReferences && el is PsiReference) {
            el = el.resolve() ?: return null
        }

        if (el is T) {
            return el
        }

        if (el is PsiFile || el is PsiDirectory || stop(el)) {
            return null
        }

        el = el.parent ?: return null
    }
}

// Children
@Contract(pure = true)
fun PsiClass.findFirstMember(): PsiMember? = findChild()

@Contract(pure = true)
fun PsiElement.findNextMember(): PsiMember? = findSibling(true)

@Contract(pure = true)
private inline fun <reified T : PsiElement> PsiElement.findChild(): T? {
    return firstChild?.findSibling(strict = false)
}

@Contract(pure = true)
private inline fun <reified T : PsiElement> PsiElement.findSibling(strict: Boolean): T? {
    var sibling = if (strict) nextSibling ?: return null else this
    while (true) {
        if (sibling is T) {
            return sibling
        }

        sibling = sibling.nextSibling ?: return null
    }
}

@Contract(pure = true)
inline fun PsiElement.findLastChild(condition: (PsiElement) -> Boolean): PsiElement? {
    var child = firstChild ?: return null
    var lastChild: PsiElement? = null

    while (true) {
        if (condition(child)) {
            lastChild = child
        }

        child = child.nextSibling ?: return lastChild
    }
}

@Contract(pure = true)
fun <T : Any> Stream<T>.filter(filter: ElementFilter?, context: PsiElement): Stream<T> {
    filter ?: return this
    return filter { filter.isClassAcceptable(it.javaClass) && filter.isAcceptable(it, context) }
}

@Contract(pure = true)
fun Stream<out PsiElement>.toResolveResults(): Array<ResolveResult> = map(::PsiElementResolveResult).toTypedArray()

fun PsiParameterList.synchronize(newParams: List<PsiParameter>) {
    ChangeSignatureUtil.synchronizeList(this, newParams, {it.parameters.asList()}, BooleanArray(newParams.size))
}

@get:Contract(pure = true)
val PsiElement.constantValue: Any?
    get() = JavaPsiFacade.getInstance(project).constantEvaluationHelper.computeConstantExpression(this)

@get:Contract(pure = true)
val PsiElement.constantStringValue: String?
    get() = constantValue as? String

@Contract(pure = true)
fun PsiType.isErasureEquivalentTo(other: PsiType): Boolean {
    // TODO: Do more checks for generics instead
    return TypeConversionUtil.erasure(this) == TypeConversionUtil.erasure(other)
}
